### DISCUSSION TITLE
Tasmota driver for custom board

### DISCUSSION THREAD

**Discussion by davidk88** [arendst/Tasmota#13673] [Q&A] [answered]
Hello,

I have some board with relays and inputs for buttons. It has own controller on it, which receives commands for relays and reports input status via uart. I would like to write a driver for tasmota to communicate with this board to set relays and for reading inputs. Is it possible to use tasmota features like shuters and blinds, but relay control routed via uart (via my driver). And also if it is possible to read input status via uart and use tasmota multipress logic, rules triggers, etc.

Best regards

**Comment by barbudor** [comment, marked as answer]
Hi
You can implement POWER control through your MCU, so any POWER command reaching Tasmota by MQTT, Console or even Web button will send the appropriate message to your MCU. This includes supporting Shutters.

Regarding button multi-press, I don't think it is possible to push from the driver multi-press events to Tasmota in a way that Tasmota will handle those as native. 
However you can still push those those as JSON events and use Rules on Tasmota sides to capture them.

Start here : https://tasmota.github.io/docs/For-Developers/

I see a few drivers that are using the TasmotaSerial class that could be suitable:
- xdrv_44_miel_hvac
- xdrv_45_shelly_dmmer

Regarding how Relays that are not directly connected to the ESP can be managed as POWER devices, have a look at drivers for PCF8574 and MCP270xx I/O expanders.

These are quite large drivers but you should see
- how they are initialized and can handle POWER  commands
- how they handle serial comm

There are many other drivers that use TasmotaSerial but mainly those are sensors or energy monitoring chip/devices so not really the best representative case.

Do not hesitate to ask for more here.

**Reply by davidk88** [reply]
Thank you for response.

I will dig in and ask here if I will have more questions.

**Comment by phpcoder** [comment]
I have made a [few examples](https://github.com/phpcoder/tasmota-custom-integration) of custom drivers. Hope it will be helpful.
